frame/ethereum/src/lib.rs

```diff
@@ -148,14 +150,20 @@ where
 		}
 	}
 
-	pub fn validate_self_contained(&self, origin: &H160) -> Option<TransactionValidity> {
+	pub fn validate_self_contained(
+		&self,
+		origin: &H160,
+		dispatch_info: &DispatchInfoOf<T::Call>,
+		len: usize,
+	) -> TransactionValidity {
 		if let Call::transact { transaction } = self {
-			Some(Pallet::<T>::validate_transaction_in_pool(
-				*origin,
-				transaction,
-			))
+			let weight_validation = CheckWeight::<T>::do_validate(dispatch_info, len)?;
+			let pool_validation = Pallet::<T>::validate_transaction_in_pool(*origin, transaction)?;
+			Ok(weight_validation.combine_with(pool_validation))
 		} else {
-			None
+			Err(TransactionValidityError::Invalid(
```

Comment:
Just keep return None here. This is to indicate that for the current path we don't know.

It'll be turned to invalid transaction if all other paths also return None.